File: JS/BOJ-S1-2583-Calculating_area.js
'use strict';

const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : '../input.txt';
let input = fs
  .readFileSync(filePath)
  .toString()
  .trim()
  .split('\n')
  .map((v) => v.trimEnd());

let output = [];
const solution = (input) => {
  const [M, N, K] = input[0].split(' ').map(Number);
  const board = Array.from({ length: M }, () => Array(N).fill(0));
  const visited = Array.from({ length: M }, () => Array(N).fill(0));
  const dx = [0, 0, -1, 1];
  const dy = [-1, 1, 0, 0];

  const ret = [];

  const area = input.slice(1).map((v) => v.split(' ').map(Number));
  area.forEach((v) => {
    const [x1, y1, x2, y2] = v;
    for (let i = y1; i < y2; i++) {
      for (let j = x1; j < x2; j++) {
        board[i][j] = 1;
      }
    }
  });

  const dfs = (y, x) => {
    visited[y][x] = 1;
    let cnt = 1;
    for (let i = 0; i < 4; i++) {
      const ny = y + dy[i];
      const nx = x + dx[i];
      if (ny < 0 || ny >= M || nx < 0 || nx >= N) continue;
      if (visited[ny][nx] === 1 || board[ny][nx] === 1) continue;
      cnt += dfs(ny, nx);
    }
    return cnt;
  };

  for (let i = 0; i < M; i++) {
    for (let j = 0; j < N; j++) {
      if (visited[i][j] === 0 && board[i][j] === 0) {
        ret.push(dfs(i, j));
      }
    }
  }
  ret.sort((a, b) => a - b);

  output.push(ret.length, '\n');
  output.push(ret.join(' '));
};

solution(input);
fs.writeFileSync('/dev/stdout', output.join('').trimEnd());
// fs.writeFileSync(
//   '../output.txt',
//   output.join('').trimEnd(),
//   (err) => console.error
// );

File: CPP/BOJ-S1-2583-Calculating_area.cpp
#include <bits/stdc++.h>
using namespace std;

#define y1 aaaa

int m, n, k, x1, y1, x2, y2;
int a[102][102], visited[102][102];
int dy[4] = {-1, 0, 1, 0};
int dx[4] = {0, 1, 0, -1};
vector<int> ret;

int dfs(int y, int x)
{
  visited[y][x] = 1;
  int ret = 1;

  for (int i = 0; i < 4; ++i)
  {
    int ny = y + dy[i];
    int nx = x + dx[i];

    if (ny < 0 || ny >= m || nx < 0 || nx >= n)
      continue;
    if (a[ny][nx] == 1 || visited[ny][nx] == 1)
      continue;

    ret += dfs(ny, nx);
  }

  return ret;
}

int main()
{
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);
  cout.tie(NULL);
  // freopen("input.txt", "r", stdin);
  // freopen("output.txt", "w", stdout);

  cin >> m >> n >> k;
  for (int i = 0; i < k; ++i)
  {
    cin >> x1 >> y1 >> x2 >> y2;
    for (int x = x1; x < x2; ++x)
    {
      for (int y = y1; y < y2; ++y)
      {
        a[y][x] = 1;
      }
    }
  }

  for (int i = 0; i < m; ++i)
  {
    for (int j = 0; j < n; ++j)
    {
      if (a[i][j] != 1 && visited[i][j] == 0)
      {
        ret.push_back(dfs(i, j));
      }
    }
  }

  sort(ret.begin(), ret.end());
  cout << ret.size() << "\n";
  for (int a : ret)
  {
    cout << a << " ";
  }

  return 0;
}